poly/src/evaluations/multivariate/multilinear/sparse.rs

```diff
@@ -0,0 +1,589 @@
+//! multilinear polynomial represented in sparse evaluation form.
+
+use crate::evaluations::multivariate::multilinear::swap_bits;
+use crate::{DenseMultilinearExtension, MultilinearExtension};
+use ark_ff::{Field, Zero};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, Read, SerializationError, Write};
+use ark_std::collections::BTreeMap;
+use ark_std::fmt::{Debug, Formatter};
+use ark_std::iter::FromIterator;
+use ark_std::ops::{Add, AddAssign, Index, Neg, Sub, SubAssign};
+use ark_std::vec::Vec;
+use ark_std::{fmt, UniformRand};
+use hashbrown::HashMap;
+use rand::Rng;
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+/// Stores a multilinear polynomial in sparse evaluation form.
+#[derive(Clone, PartialEq, Eq, Hash, Default, CanonicalSerialize, CanonicalDeserialize)]
+pub struct SparseMultilinearExtension<F: Field> {
+    /// tuples of index and value
+    pub evaluations: BTreeMap<usize, F>,
+    /// number of variables
+    pub num_vars: usize,
+    zero: F,
```

Comment:
Why is this needed?

Reply:
`index` requires returning a reference, and the sparse list (i.e. the TreeMap) only contains nonzero elements, so `zero` serves as a dummy variable (see https://github.com/arkworks-rs/algebra/blob/366749faee1321c791b1a91ffa736e1bcff82aa6/poly/src/evaluations/multivariate/multilinear/sparse.rs#L209). Is there any better way?

Reply:
I see. This is fine =)